src/client/components/cube-view/cube-view.tsx

```diff
@@ -149,18 +151,23 @@ export class CubeView extends React.Component<CubeViewProps, CubeViewState> {
   }
 
   componentWillReceiveProps(nextProps: CubeViewProps) {
-    var { essence } = nextProps;
-    if (!essence.equals(this.state.essence)) {
-      this.setState({ essence });
+    var hashEssence = this.getEssenceFromHash(nextProps.hash);
+    if (hashEssence && !hashEssence.equals(this.state.essence)) {
+      this.setState({ essence: hashEssence });
+    }
+
+    if (!this.props.selectedDataSource.equals(nextProps.selectedDataSource)) {
+      var newEssence = Essence.fromDataSource(nextProps.selectedDataSource, {
+        dataSource: nextProps.selectedDataSource,
+        visualizations
+      });
+      this.setState({ essence: newEssence });
```

Comment:
There are two `this.setState({ essence: ... })` calls here that both can execute. I feel like it would be clearer if there is only one path to call setState